### PR TITLE
enhance: add bitmap offset cache to speed up retrieve raw data

### PR DESCRIPTION
**Pull request by zhagnlu**
#35458

### CODE REVIEW

File: internal/core/src/index/BitmapIndex.cpp [hunk 1]
```diff
@@ -726,24 +750,10 @@ BitmapIndex<T>::Reverse_Lookup(size_t idx) const {
     AssertInfo(idx < total_num_rows_, "out of range of total coun");
 
     if (build_mode_ == BitmapIndexBuildMode::ROARING) {
-        for (auto it = data_.begin(); it != data_.end(); it++) {
-            for (const auto& v : it->second) {
```
Comment (xiaofan-luan) [inline internal/core/src/index/BitmapIndex.cpp:730]: if we use it->second.contais(), could it be faster?

File: internal/core/src/index/BitmapIndex.cpp [hunk 2]
```diff
@@ -333,6 +333,29 @@ BitmapIndex<T>::DeserializeIndexData(const uint8_t* data_ptr,
     }
 }
 
+template <typename T>
+void
+BitmapIndex<T>::BuildOffsetCache() {
+    if (build_mode_ == BitmapIndexBuildMode::ROARING) {
+        data_offsets_cache_.resize(total_num_rows_);
+        for (auto it = data_.begin(); it != data_.end(); it++) {
```
Comment (xiaofan-luan) [inline internal/core/src/index/BitmapIndex.cpp:341]: this might takes too much memory ?

**Comment by xiaofan-luan** [comment]
I don't think we need this cache to be enabled by default.
Probably takes too much memory? if we have it we can close it by a config by default

**Comment by zhagnlu** [comment]
> I don't think we need this cache to be enabled by default. Probably takes too much memory? if we have it we can close it by a config by default

ok，actually it cost 8 * rows bytes for a segment, 100M segment cost 8M memory.  for newest code,  I add a config that allow user to open or close it by alter_index. default is closed

File: internal/datacoord/index_service.go [hunk 3]
```diff
@@ -272,13 +272,19 @@ func (s *Server) CreateIndex(ctx context.Context, req *indexpb.CreateIndexReques
...
 		return merr.WrapErrParameterInvalidMsg("invalid mmap index params", err.Error())
 	}
-	userIndexParams := funcutil.KeyValuePair2Map(index.UserIndexParams)
 	if err := indexparamcheck.ValidateMmapIndexParams(indexType, userIndexParams); err != nil {
 		return merr.WrapErrParameterInvalidMsg("invalid mmap user index params", err.Error())
 	}
+	if err := indexparamcheck.ValidateOffsetCacheIndexParams(indexType, indexParams); err != nil {
+		return merr.WrapErrParameterInvalidMsg("invalid offset cache index params", err.Error())
+	}
+	if err := indexparamcheck.ValidateOffsetCacheIndexParams(indexType, userIndexParams); err != nil {
```
Comment (xiaofan-luan) [inline internal/datacoord/index_service.go:285]: seems this is duplicated

**Comment by xiaofan-luan** [comment]
except the duplicate, everthing looks good to me

@zhagnlu 

/approve

**Comment by zhagnlu** [comment]
> except the duplicate, everthing looks good to me
> 
> @zhagnlu
> 
> /approve

not repeat, indexPram and userIndexParam check

**Pull request merged**